lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -313,95 +391,152 @@ def _is_tls_fully_configured(self) -> bool:
+        if current_update_node_conf:
+            self.peers_data.put_object(
+                Scope.UNIT, "last-update-node-config", current_update_node_conf
+            )
+
+    def _stop_opensearch(self) -> None:
         """Stop OpenSearch if possible."""
-        try:
-            self.unit.status = WaitingStatus(ServiceIsStopping)
-            self.opensearch.stop()
-            self.unit.status = WaitingStatus(ServiceStopped)
-        except OpenSearchStopError as e:
-            logger.debug(e)
-            self.unit.status = BlockedStatus(ServiceStopFailed)
-            event.defer()
+        if not self.opensearch.is_started():
+            return
+
+        self.unit.status = WaitingStatus(ServiceIsStopping)
+
+        # 1. Add current node to the voting + alloc exclusions
+        self.opensearch_exclusions.add_current()
+
+        # 2. stop the service
+        self.opensearch.stop()
+        self.unit.status = WaitingStatus(ServiceStopped)
+
+        # 3. Remove the exclusions
+        self.opensearch_exclusions.delete_current()
 
     def _restart_opensearch(self, event: EventBase) -> None:
         """Restart OpenSearch if possible."""
-        self._stop_opensearch(event)
+        if not self.peers_data.get(Scope.UNIT, "starting", False):
+            try:
+                self._stop_opensearch()
+            except OpenSearchStopError as e:
+                logger.error(e)
+                event.defer()
+                return
+
         self._start_opensearch(event)
 
-    def _can_service_start(self):
+    def _can_service_start(self) -> bool:
         """Return if the opensearch service can start."""
         # if there are any missing system requirements leave
         missing_sys_reqs = self.opensearch.missing_sys_requirements()
         if len(missing_sys_reqs) > 0:
             self.unit.status = BlockedStatus(" - ".join(missing_sys_reqs))
             return False
 
+        # if self.peers_data.get(Scope.APP, "unit-busy-starting", False):
+        #     return False
+
         # When a new unit joins, replica shards are automatically added to it. In order to prevent
         # overloading the cluster, units must be started one at a time. So we defer starting
         # opensearch until all shards in other units are in a "started" or "unassigned" state.
-        if not self.unit.is_leader() and self.peers_data.get(Scope.APP, "leader_ip"):
-            try:
-                busy_shards = ClusterState.busy_shards_by_unit(
-                    self.opensearch, self.peers_data.get(Scope.APP, "leader_ip")
-                )
-                if busy_shards:
-                    message = WaitingForBusyShards.format(
-                        " - ".join([f"{key}/{','.join(val)}" for key, val in busy_shards.items()])
-                    )
-                    self.unit.status = WaitingStatus(message)
-                    return False
+        if not self.unit.is_leader():
```

Comment:
*Suggestion:* There's a lot of layers of indentation in this function - consider the following simplifications:
- returning `True` early `if self.unit.is_leader()`, 
- returning `True` early `if not self.peers_data.get(Scope.APP, "leader_ip")`,